pkg/controllers/user/alert/configsyncer/configsyncer.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func NewConfigSyncer(cluster *config.UserContext, manager *manager.Manager) *ConfigSyncer {
+	return &ConfigSyncer{
+		secretClient:       cluster.Core.Secrets("cattle-alerting"),
+		clusterAlertLister: cluster.Management.Management.ClusterAlerts(cluster.ClusterName).Controller().Lister(),
+		projectAlertLister: cluster.Management.Management.ProjectAlerts("").Controller().Lister(),
+		notifierLister:     cluster.Management.Management.Notifiers(cluster.ClusterName).Controller().Lister(),
+		clusterName:        cluster.ClusterName,
+		alertManager:       manager,
+	}
+
+}
+
+type ConfigSyncer struct {
+	secretClient       v1.SecretInterface
+	projectAlertLister v3.ProjectAlertLister
+	clusterAlertLister v3.ClusterAlertLister
+	notifierLister     v3.NotifierLister
+	clusterName        string
+	alertManager       *manager.Manager
+}
+
+func (d *ConfigSyncer) ProjectSync(key string, alert *v3.ProjectAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) ClusterSync(key string, alert *v3.ClusterAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) NotifierSync(key string, alert *v3.Notifier) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) sync() error {
+	//logrus.Info("start sync config")
+
+	clusterAlerts, err := d.clusterAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating cluster alerts")
+	}
+
+	projectAlerts, err := d.projectAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating project alerts")
+	}
+
+	pAlerts := []*v3.ProjectAlert{}
+	for _, alert := range projectAlerts {
+		if controller.ObjectInCluster(d.clusterName, alert) {
+			pAlerts = append(pAlerts, alert)
+		}
+	}
+
+	notifiers, err := d.notifierLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating project alerts")
+	}
+
+	config := d.alertManager.GetDefaultConfig()
+	config.Global.PagerdutyURL = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+	d.addClusterAlert2Config(config, clusterAlerts, notifiers)
+	d.addProjectAlert2Config(config, pAlerts, notifiers)
+
+	data, err := yaml.Marshal(config)
+	//logrus.Infof("after updating notifier: %s", string(data))
+	if err != nil {
+		return errors.Wrapf(err, "Marshal secrets")
+	}
+
+	configSecret, err := d.secretClient.Get("alertmanager", metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "Get secrets")
+	}
+
+	configSecret.Data["config.yml"] = data
+	_, err = d.secretClient.Update(configSecret)
+	if err != nil {
+		return errors.Wrapf(err, "Update secrets")
+	}
+
+	go func() {
+		for i := 0; i < 5; i++ {
+			d.alertManager.ReloadConfiguration()
```

Comment:
why are we reloading configuration 5 times even if no error happens? Looks like ReloadConfiguration can return an error

Reply:
It is because I find after the secret is updated, the config file in the alertmanager container isn't up to date for a while,  even I can get the updated content by using `kubectl get secret`.  
 It takes a while for the config file in the container to reflect the change.  What I do is to reload the config several times to make sure alert manager can load the new config. Do you have any suggestions?